tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -627,23 +627,59 @@ class ToEtcExpr(annotationStore: AnnotationStore, varPool: TypeVarPool) extends
           accessorsWithNewValues.flatMap(p => List(this(p._1), this(p._2)))
         mkApp(ref, disjunctiveType, this(fun) +: xargs: _*)
 
-      case OperEx(TlaFunOper.recFunDef, body, NameEx(name), bindingSet) =>
-        // the expected type is: (((a -> b) => (a => b)) => (a -> b)) (λ $recFun ∈ Set(c -> d). λ x ∈ Int. x)
+      case OperEx(TlaFunOper.recFunDef, body, args @ _*) =>
+        // fun[x \in S, y \in T |-> e] == ...
+        // or, fun[<<x, y>> \in S, z \in T |-> e] == ...
+        //
+        // The expected type is for a one-argument function is:
+        // (((b -> a) => (b => a)) => (b -> a)) (λ $recFun ∈ Set(c -> d). λ x ∈ Int. x)
+        //
+        // (The case of multiple arguments is quite complex)
+        val bindings =
+          translateBindings(
+              args
+                .grouped(2)
+                .map {
+                  case Seq(varEx, setEx) => (varEx, setEx)
+                  case orphan            => throw new TypingException(s"Invalid bound variables and sets ${orphan} in: ${ex}")
+                }
+                .toSeq: _*
+          )
+
         val a = varPool.fresh
-        val b = varPool.fresh
-        val funType = FunT1(a, b)
-        val aToB = OperT1(Seq(a), b)
-        val principal = OperT1(Seq(OperT1(Seq(funType), aToB)), funType)
+        // start with "b", as "a" goes to the result
```

Comment:
I'm afraid this this comment might not be very clear to other readers. It's not at all obvious how are we starting with `b`, since `b` doesn't appear here.

Should `a` perhaps be called `resultType`? That would make half this comment unnecessary.

Reply:
that sounds good :)